env/fs_on_demand.cc

```diff
@@ -0,0 +1,344 @@
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "env/fs_on_demand.h"
+
+#include <set>
+
+#include "file/filename.h"
+#include "rocksdb/types.h"
+
+namespace ROCKSDB_NAMESPACE {
+// Check if the input path is under the destination directory, and if so,
+// change it to the equivalent path in the source directory.
+// Return value is true if the path was modified, false otherwise
+bool OnDemandFileSystem::CheckPathAndAdjust(std::string& path) {
+  size_t pos = path.find(dest_path_);
+  if (pos > 0) {
+    return false;
+  }
+  path.replace(pos, dest_path_.length(), src_path_);
+  return true;
+}
+
+bool OnDemandFileSystem::LookupFileType(const std::string& name,
+                                        FileType* type) {
+  std::size_t found = name.find_last_of('/');
+  std::string file_name = name.substr(found);
+  uint64_t number = 0;
+  return ParseFileName(file_name, &number, type);
+}
+
+// RocksDB opens non-SST files for reading in sequential file mode. This
+// includes CURRENT, OPTIONS, MANIFEST etc. For these files, we open them
+// in place in the source directory. For files that are appendable or
+// can be renamed, which is MANIFEST and CURRENT files, we wrap the
+// underlying FSSequentialFile with another class that checks when EOF
+// has been reached and re-opens the file to see the latest data. On some
+// distributed file systems, this is necessary.
+IOStatus OnDemandFileSystem::NewSequentialFile(
+    const std::string& fname, const FileOptions& file_opts,
+    std::unique_ptr<FSSequentialFile>* result, IODebugContext* dbg) {
+  FileType type;
+  static std::set<FileType> valid_types(
+      {kWalFile, kDescriptorFile, kCurrentFile, kIdentityFile, kOptionsFile});
+  if (!LookupFileType(fname, &type) ||
+      (valid_types.find(type) == valid_types.end())) {
+    return IOStatus::NotSupported();
+  }
+
+  IOStatus s;
+  std::string rname = fname;
+  if (CheckPathAndAdjust(rname)) {
+    // First clear any local directory cache as it may be out of date
+    target()->DiscardCacheForDirectory(rname);
+
+    s = target()->FileExists(fname, file_opts.io_options, dbg);
+    if ((type == kDescriptorFile || type == kCurrentFile) && s.ok()) {
+      s = target()->DeleteFile(fname, file_opts.io_options, nullptr);
+      if (s.ok()) {
+        s = IOStatus::NotFound();
+      }
+    }
+    if (s.IsNotFound() || s.IsPathNotFound()) {
+      std::unique_ptr<FSSequentialFile> inner_file;
+      s = target()->NewSequentialFile(rname, file_opts, &inner_file, dbg);
+      if (s.ok()) {
+        result->reset(new OnDemandSequentialFile(std::move(inner_file), this,
+                                                 file_opts, rname));
+      }
+      return s;
+    }
+  }
+
+  if (s.ok() || s.IsNotFound() || s.IsPathNotFound()) {
+    s = target()->NewSequentialFile(fname, file_opts, result, dbg);
+  }
+  return s;
+}
+
+// This is only supported for SST files. If the file is present locally,
```

Comment:
At least for the new BlobDB, where blob files are immutable, the same thing could be done for blob files as for SSTs.

Reply:
Yeah, it should be trivial to support blob files as well. We might want to sanity check in `DB::OpenAsFollower` that its the new BlobDB and fail otherwise. I'll keep a note of it and address it in a follow-on PR.